### PR TITLE
Calculate maneuver times and angle from cmds, replace maneuver summary file

### LINKED ISSUE #409: Finish starcheck changes to deal with observations with MANSTART
- [ ] https://github.com/sot/starcheck/pull/408

### PR DESCRIPTION
**Pull request by jeanconn**
## Description
Calculate maneuvers from backstop cmds and replace maneuver summary as source of maneuver times and angle.

This fixes the known issue that the maneuver summary file is incorrect for observations with nonzero MANSTART.
Fixes #409 

## Interface impacts

Starcheck text output changed slightly, no impact to parsing tools:
- Maneuver end time changed by 10 seconds (new value is correct)

## Testing
Tested with ska3-matlab 2023.4rc6 on fido, though for these changes no special test setup should be required.

### Unit tests

- [x] No unit tests

### Functional tests

Used the "new" regression test set we used to review https://github.com/sot/starcheck/pull/402 plus two weeks with MANSTART values that caused issues with maneuver end times in flight starcheck
```
jeanconn-fido> more run.sh
jeanconn-fido> more run.sh
# Long week and IR Zone holds
/home/jeanconn/git/starcheck/sandbox_starcheck -dir /data/mpcrit1/mplogs/2022/DEC2622/oflsa -out dec2622a_test
/proj/sot/ska3/flight/bin/starcheck -dir /data/mpcrit1/mplogs/2022/DEC2622/oflsa -out dec2622a_flight
/proj/sot/ska/bin/diff2html dec2622a_flight.txt dec2622a_test.txt > dec2622a_diff.html

# Monitor star
/home/jeanconn/git/starcheck/sandbox_starcheck -dir /data/mpcrit1/mplogs/2022/NOV2822/oflsa -out nov2822a_test
/proj/sot/ska3/flight/bin/starcheck -dir /data/mpcrit1/mplogs/2022/NOV2822/oflsa -out nov2822a_flight
/proj/sot/ska/bin/diff2html nov2822a_flight.txt nov2822a_test.txt > nov2822a_diff.html

# Maneuver only loads
/home/jeanconn/git/starcheck/sandbox_starcheck -dir /data/mpcrit1/mplogs/2022/OCT2422/oflsb -out oct2422b_test
/proj/sot/ska3/flight/bin/starcheck -dir /data/mpcrit1/mplogs/2022/OCT2422/oflsb -out oct2422b_flight
/proj/sot/ska/bin/diff2html oct2422b_flight.txt oct2422b_test.txt > oct2422b_diff.html

# Replan
/home/jeanconn/git/starcheck/sandbox_starcheck -dir /data/mpcrit1/mplogs/2021/MAY0521/oflsa -out may0521a_test
/proj/sot/ska3/flight/bin/starcheck -dir /data/mpcrit1/mplogs/2021/MAY0521/oflsa -out may0521a_flight
/proj/sot/ska/bin/diff2html may0521a_flight.txt may0521a_test.txt > may0521a_diff.html

# MANSTART in APR0323A
/home/jeanconn/git/starcheck/sandbox_starcheck -dir /data/mpcrit1/mplogs/2023/APR0323/oflsa -out apr0323a_test
/proj/sot/ska3/flight/bin/starcheck -dir /data/mpcrit1/mplogs/2023/APR0323/oflsa -out apr0323a_flight
/proj/sot/ska/bin/diff2html apr0323a_flight.txt apr0323a_test.txt > apr0323a_diff.html

# MANSTART in APR1723A
/home/jeanconn/git/starcheck/sandbox_starcheck -dir /data/mpcrit1/mplogs/2023/APR1723/oflsa -out apr1723a_test
/proj/sot/ska3/flight/bin/starcheck -dir /data/mpcrit1/mplogs/2023/APR1723/oflsa -out apr1723a_flight
/proj/sot/ska/bin/diff2html apr1723a_flight.txt apr1723a_test.txt > apr1723a_diff.html

# JUN1923 week big dither
/home/jeanconn/git/starcheck/sandbox_starcheck -dir /data/mpcrit1/mplogs/2023/JUN1923/oflsa -out jun1923a_test
/proj/sot/ska3/flight/bin/starcheck -dir /data/mpcrit1/mplogs/2023/JUN1923/oflsa -out jun1923a_flight
/proj/sot/ska/bin/diff2html jun1923a_flight.txt jun1923a_test.txt > jun1923a_diff.html
```
Output in https://icxc.cfa.harvard.edu/aspect/test_review_outputs/starcheck-pr408/

#### Test Review

@taldcroft looked through all the diffs and found only expected changes.

### CODE REVIEW

**Comment by jeanconn** [comment]
This strategy is just to reuse the maneuver calculation code that was being called in hopper and fiddle with that data structure.  An alternative path would be to just make a similar structure with kadi.commands, or reorganize more significantly.

File: starcheck/src/lib/Ska/Starcheck/Obsid.pm [hunk 1]
```diff
@@ -372,39 +370,12 @@ sub set_maneuver {
...
-                my @q_man = @{ $q_man->{q} };
-                my $q_diff = $q_man->divide($q_obc);
 
-                if (   abs($q_diff->{ra0} * 3600) > 1.0
-                    || abs($q_diff->{dec} * 3600) > 1.0
-                    || abs($q_diff->{roll0} * 3600) > 10.0)
-                {
-                    push @{ $self->{warn} },
-                      sprintf(
-"Target uplink precision problem for MP_TARGQUAT at $c->{date}\n"
```
Comment (jeanconn) [inline starcheck/src/lib/Ska/Starcheck/Obsid.pm:394]: This PR removes the uplink precision check as I'm not sure what to check if we don't read the maneuver summary for comparison.

**Comment by jeanconn** [comment]
@taldcroft Do you think it makes sense to use the hopper maneuver code for this?  Or would it make more sense to retire hopper in general as a nice idea but mostly more than we needed and use some new maneuver code built with kadi.commands and replace the hopper attitude checking code with something simpler?

File: starcheck/src/lib/Ska/Starcheck/Obsid.pm (hunk 1 above)
Comment (taldcroft) [reply starcheck/src/lib/Ska/Starcheck/Obsid.pm:394]: I'm fine with removing the uplink precision check.

**Comment by taldcroft** [comment]
Using hopper is fine for now.

**Comment by taldcroft** [comment]
Sorry for going back and forth, but I looked at the code and again thought there must be a better way. I managed to find the old thread "Issue with MCC loading maneuver summary for AUG1312A" with Eric's summary of the flight calculation:
```
     Q_Mag := ( Q_Targ_Cmd(1)*Q_Targ_Cmd(1) ) +
             ( Q_Targ_Cmd(2)*Q_Targ_Cmd(2) ) +
             ( Q_Targ_Cmd(3)*Q_Targ_Cmd(3) );

     Q_Targ_Cmd(4) := SQRT( ABS(1.0 - Q_Mag ));  # Edit to include the ABS() there
     QUATERNION_NORMALIZE( Q_Targ_Cmd, Q_Targ_Cmd, Q_Norm_OK );
```
So why not have starcheck use exactly the OBC-normalized quaternion from above. Then there is no need to check for consistency because we are using the right answer. We could leave in the check about normalization, but I ran that check on the entire mission with kadi commands and it has never failed. Backstop just isn't going to deliver badly-normalized quaternions and we rely on those all over the place without sanity checks.

**Comment by taldcroft** [comment]
In theory we could add a kwarg to the `Quat` class and `Quaternion.normalize` to tell it to normalize the input like the OBC does.

**Comment by jeanconn** [comment]
"So why not have starcheck use exactly the OBC-normalized quaternion from above."  That's basically what I was thinking but I was leaning toward doing it in a future release.  The issues is that as you noted in your on-the-side review, what starcheck is using seems to not really be nailed down... it seems to use something different for reference attitude in each method, so I thought it would be more of a little project to review and make uniform.  But we could do this now.

Though with regard to your kadi cmds review --- that was of approved products and not including early mission.

**Comment by taldcroft** [comment]
OK, we can defer making improvements in the way starcheck works to later. I'll approve this once you summarize the outcome of the functional testing in the description.

File: starcheck/pcad_att_check.py [hunk 2]
```diff
@@ -139,7 +139,42 @@ def make_pcad_attitude_check_report(backstop_file, or_list_file=None, attitude_f
...
+    mm = []
+    for m in sc.maneuvers:
+        q1 = Quat(q=Quaternion.normalize(
+            [m['initial']['q1'], m['initial']['q2'],
+             m['initial']['q3'], m['initial']['q4']]))
+        q2 = Quat(q=Quaternion.normalize(
+            [m['final']['q1'], m['final']['q2'],
+             m['final']['q3'], m['final']['q4']]))
+
+        angle = np.degrees(2 * np.arccos(q1.q.dot(q2.q)))
```
Comment (jeanconn) [inline starcheck/pcad_att_check.py:153]: One thing I learned in quick functional test output review is that the maneuver angle we have been using from the matlab tools is not the great circle angle (sphere_dist) and is instead more accurately reflected as this (not sure what you call this angle - rotation matrix angle?).  I don't know if that is actually interesting at all for acquisition probability.

File: starcheck/pcad_att_check.py [hunk 3]
```diff
@@ -1,10 +1,13 @@
 # Licensed under a 3-clause BSD style license - see LICENSE.rst
 import re
 from astropy.table import Table
+import numpy as np
 import Quaternion
+from Quaternion import Quat
 
 from parse_cm import read_backstop, read_or_list
 from Chandra.Time import DateTime
+from agasc import sphere_dist
```
Comment (jeanconn) [inline starcheck/pcad_att_check.py:10]: Forgot to cut this.

File: starcheck/pcad_att_check.py (hunk 2 above)
Comment (jeanconn) [reply starcheck/pcad_att_check.py:153]: For this maneyver from DEC2622A, the maneuver from 44926 to 44925 has a maneuver 'angle' of 165.05 degrees, but a sphere_dist of 14.6 deg.

```
OBSID: 44926  
RA, Dec, Roll (deg):    93.000000    23.000000   271.874615
Dither: ON Y_amp= 8.0  Z_amp= 8.0  Y_period=1000.0  Z_period= 707.1 
BACKSTOP GUIDE_SUMM OR DOT TLR 

MP_TARGQUAT at 2023:001:10:27:59.567 (VCDU count = 7072498)
  Q1,Q2,Q3,Q4: -0.36518414  -0.59294570  0.41533499  0.58528010
  MANVR: Angle=  74.61 deg  Duration= 1655 sec  Slew err= 51.3 arcsec  End= 2023:001:10:55:40

==================================================================================== 
OBSID: 44925  
RA, Dec, Roll (deg):   108.597458    20.909120    70.938652
Dither: ON Y_amp= 8.0  Z_amp= 8.0  Y_period=1000.0  Z_period= 707.1 
BACKSTOP GUIDE_SUMM OR DOT TLR 

MP_TARGQUAT at 2023:001:11:50:04.357 (VCDU count = 7091716)
  Q1,Q2,Q3,Q4: 0.45300842  0.37715544  0.71183683  0.38187101
  MANVR: Angle= 165.05 deg  Duration= 2861 sec  Slew err= 27.6 arcsec  End= 2023:001:12:37:50

```

Comment (taldcroft) [reply starcheck/pcad_att_check.py:153]: This is a great point and one I hope we'll remember. Fundamentally the way I would think about this is:
```
dq = q1.dq(q2)  # quat getting from q1 to q2 (or vice versa, I never remember)
Then dq.q[3] = cos(2 * alpha) where alpha is the rotation angle about q[0:2].
I'm assuming that q1.q .dot (q2.q) is the same as alpha?

Q56. How do I convert a rotation axis and angle to a quaternion?
----------------------------------------------------------------

  Given the rotation axis and angle, the following algorithm may be
  used to generate a quaternion:

    ------------------------------------------------
    sin_a = sin( angle / 2 )
    cos_a = cos( angle / 2 )

    q -> x    = axis -> x * sin_a
    q -> y    = axis -> y * sin_a
    q -> z    = axis -> z * sin_a
    q -> w    = cos_a

    quaternion_normalise( q );
    ------------------------------------------------
```
I almost wonder if we should add a method to `Quat` like `Quat.maneuver_angle(self, q2)`.

**Comment by jeanconn** [comment]
Another side effect of this process is that all maneuvers appear to end 10 seconds later than the maneuver summary had previously reported.  I don't actually know which one is a closer estimate.  With the new estimate, the monitor window checks in the nov2822 test products didn't pass.  So need to figure out if we update the monitor window timing checks or subtract 10 seconds from all of the maneuver end times in this PR.

**Comment by taldcroft** [comment]
That's the standard 10 sec delta in the man summary right? The backstop timing is the truth which we should use, so updating the mon window timing would seem like the way to go.

**Comment by jeanconn** [comment]
I just wasn't sure which was actually right as there is no maneuver end time in backstop and we've been using the maneuver summary end time for quite a while.  It looks like the new derived times more closely match kadi manvr events, so that seems fine and agrees with your strategy.

**Comment by jeanconn** [comment]
I reran the tests and added one with big dither to go through those timing checks too (which seemed OK).

File: starcheck/pcad_att_check.py [hunk 4]
```diff
@@ -139,7 +139,46 @@ def make_pcad_attitude_check_report(backstop_file, or_list_file=None, attitude_f
...
+    mm = []
+    for m in sc.maneuvers:
+        q1 = Quat(q=Quaternion.normalize(
+            [m['initial']['q1'], m['initial']['q2'],
+             m['initial']['q3'], m['initial']['q4']]))
+        q2 = Quat(q=Quaternion.normalize(
+            [m['final']['q1'], m['final']['q2'],
+             m['final']['q3'], m['final']['q4']]))
+
+        # Calculate maneuver angle using code borrowed from kadi
```
Comment (jeanconn) [inline starcheck/pcad_att_check.py:153]: I updated the angle calculation to just use that 4th component -- which seems to work and was already numerically stable here with the logic from kadi commands even if q2 == q1?

**Comment by taldcroft** [comment]
@jeanconn - I was filling in the interface impact and realized that we (and especially Eric) find some value in clicking through to the maneuver summary file. Even if it is no longer used for starcheck review maybe we should keep that link there?

**Comment by taldcroft** [comment]
This is great work with a very thorough job of testing! Ready for approval and merge once we decide on the maneuver summary file.

**Comment by jeanconn** [comment]
Instead of re-running testing, since @taldcroft had already reviewed those diffs in detail, I did incremental output diffs
```
jeanconn-fido> more incremental.sh
# Long week and IR Zone holds
/home/jeanconn/git/starcheck/sandbox_starcheck -dir /data/mpcrit1/mplogs/2022/DEC2622/oflsa -out dec2622a_test2
diff dec2622a_test.txt dec2622a_test2.txt > dec2622a_diff2.txt

# Monitor star
/home/jeanconn/git/starcheck/sandbox_starcheck -dir /data/mpcrit1/mplogs/2022/NOV2822/oflsa -out nov2822a_test2
diff nov2822a_test.txt nov2822a_test2.txt > nov2822a_diff2.txt


# Maneuver only loads
/home/jeanconn/git/starcheck/sandbox_starcheck -dir /data/mpcrit1/mplogs/2022/OCT2422/oflsb -out oct2422b_test2
diff oct2422b_test.txt oct2422b_test2.txt > oct2422b_diff2.txt

# Replan
/home/jeanconn/git/starcheck/sandbox_starcheck -dir /data/mpcrit1/mplogs/2021/MAY0521/oflsa -out may0521a_test2
diff may0521a_test.txt may0521a_test2.txt > may0521a_diff2.txt

# MANSTART in APR0323A
/home/jeanconn/git/starcheck/sandbox_starcheck -dir /data/mpcrit1/mplogs/2023/APR0323/oflsa -out apr0323a_test2
diff apr0323a_test.txt apr0323a_test2.txt > apr0323a_diff2.txt

# MANSTART in APR1723A
/home/jeanconn/git/starcheck/sandbox_starcheck -dir /data/mpcrit1/mplogs/2023/APR1723/oflsa -out apr1723a_test2
diff apr1723a_test.txt apr1723a_test2.txt > apr1723a_diff2.txt

# JUN1923 week big dither
/home/jeanconn/git/starcheck/sandbox_starcheck -dir /data/mpcrit1/mplogs/2023/JUN1923/oflsa -out jun1923a_test2
diff jun1923a_test.txt jun1923a_test2.txt > jun1923a_diff2.txt
```
So the new *diff2.txt files show the expected output that MANVR is back, relative to the last round of test outputs.  And inspection of a *test2.html file shows the MANVR link works.

**Review by taldcroft — APPROVED** [review]
Looks great thanks!

**Pull request merged**